Highlights reconstruction : introduce new method

aurelienpierre opened the pull request:
We merge the spirit of guided filters, wavelets multiscale processing and isotropic diffusion to introduce a new highlights reconstruction method that guides the laplacian of the R, G, B channels with the laplacian of the norm. This goes in the usual RAW highlights module.

What it does:
* restore the expected luminance of clipped regions, compared to their surround (otherwise clipped regions are recorded darker than they should),
* smoothen some sharp artifacts,
* get *some* color back,
* copy-paste details from the norm to the clipped channels.

What it does not:
* inpaint color (too difficult to identify legitimate objects to sample color from and avoid color bleeding across "edges" that may not be there anymore),
* deal with chromatic aberrations,
* Xtrans sensors.

Where it fails:
* chromatic aberrations and other sharp color issues are not properly corrected.

Notes:

It works better in conjunction with filmic reconstruction, which uses a similar method but with different priorities (smoothness first). 

Limitations:
* No XTrans, and I don't think the method can be adapted (how to do meaningful laplacians in a non-regular CFA pattern ?)
* No OpenCL.

Demo : https://discuss.pixls.us/t/guiding-laplacians-to-restore-clipped-highlights/28493

rawfiner commented:
This image shared by @TurboGit on #10716 (which was problematic for @jenshannoschwalm ) seems also problematic here:

https://drive.google.com/file/d/1XFKbgAkbLfckY-JZKrn4mP9MsxLTc59c/view?usp=sharing

aurelienpierre commented:
This image has a too large clipped region, it's beyond reconstructibility. Also, it falls under "poor photographying skills" since clipped highlights should only be primary light sources (where you can't avoid it), not 25% of the image.

rawfiner commented:
I still think the algorithm should fallback nicely in such case, inpainting everything in grey for instance. For such case where big clouds are clipped, we don't really need details, just to avoid a too strong color cast.

Also, I found a strange behavior on this image:
https://drive.google.com/file/d/1yMHMH9N_xc-GBSCbTq7m4qyRL2YirFJf/view?usp=sharing

![Capture d’écran du 2021-12-28 20-25-53](https://user-images.githubusercontent.com/34063828/147600540-473ec3a9-c690-45b5-a3f6-39b28aa7263a.png)

aurelienpierre commented:
> I still think the algorithm should fallback nicely in such case, inpainting everything in grey for instance.

It's not possible. "Grey" is possible only after a full chromatic adaptation by enforcing R = G = B. We are here before demosaicing, we have no idea if the illuminant is right or not, we know no color at that point, only sparse gradients. Degrading to grey is the responsibility of filmic reconstruction that happens after proper color profiling and adaptation, meaning it has a sane and sanitized color framework to make assumptions upon colors.

> Also, I found a strange behavior on this image:

Latest commit should deal with that. Thanks for your tests !

rawfiner commented:
> Latest commit should deal with that.
Indeed!

Another small issue, when zooming into the image in the darkroom, image is surrounded with a row and column of red or blue pixels
![Capture d’écran du 2021-12-29 11-46-52](https://user-images.githubusercontent.com/34063828/147654550-624b0f93-33e4-42bb-aa5d-73b95482db3f.png)

Also, with new commits, the results on the puffin image are worse than before, the orange parts become green whatever the feathering values.

aurelienpierre commented:
>      Another small issue, when zooming into the image in the darkroom, image is surrounded with a row and column of red or blue pixels

It's normal, I don't demosaic the first/last rows and columns yet.

> Also, with new commits, the results on the puffin image are worse than before, the orange parts become green whatever the feathering values.

What is the last commit that works here ?

rawfiner commented:
6cc7c8a520da54bea8ddda48f5dd4a63dabf79f6 (remove WB corrections) is the first bad commit

TurboGit commented:
As we are at it, why the highlight recovery module default to "clip"? Can't we use a safe default like "reconstruct in LCh"?

Thanatomanic commented:
Hi @aurelienpierre I liked to try this out, but I get compilation errors on Windows MSYS2:

```
[696/889] Building C object lib/darktable/plugins/CMakeFiles/highlights.dir/introspection_highlights.c.obj
FAILED: lib/darktable/plugins/CMakeFiles/highlights.dir/introspection_highlights.c.obj
C:\msys64\mingw64\bin\cc.exe -DAVIF_DLL -DGDK_DISABLE_DEPRECATED -DGDK_VERSION_MIN_REQUIRED=GDK_VERSION_3_22 -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_MIN_REQUIRED -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_56 -DGTK_DISABLE_DEPRECATED -DGTK_DISABLE_SINGLE_INCLUDES -DHAVE_BUILTIN_CPU_SUPPORTS -DHAVE_CONFIG_H -DHAVE_GAME -DHAVE_GMIC -DHAVE_GPHOTO2 -DHAVE_GPHOTO_25_OR_NEWER -DHAVE_GRAPHICSMAGICK -DHAVE_HTTP_SERVER -DHAVE_ICU -DHAVE_ISO_CODES -DHAVE_LENSFUN -DHAVE_LIBAVIF=1 -DHAVE_LIBEXIV2_WITH_ISOBMFF=1 -DHAVE_LIBHEIF=1 -DHAVE_LIBRAW=1 -DHAVE_LIBSECRET -DHAVE_MAP -DHAVE_OPENCL -DHAVE_OPENEXR -DHAVE_OPENJPEG -DHAVE_OSMGPSMAP_110_OR_NEWER -DHAVE_SQLITE_324_OR_NEWER -DNATIVE_ARCH -DSQLITE_CORE -DSQLITE_ENABLE_ICU -DUSE_LUA -D_POSIX_THREAD_SAFE_FUNCTIONS -D_USE_MATH_DEFINES -D__GDK_KEYSYMS_COMPAT_H__ -D__USE_MINGW_ANSI_STDIO=1 -Dhighlights_EXPORTS -IC:/msys64/home/Roel/darktable/build-hr/lib/darktable/plugins -IC:/msys64/home/Roel/darktable/src/iop -IC:/msys64/home/Roel/darktable/src -IC:/msys64/home/Roel/darktable/src/external/LuaAutoC -IC:/msys64/home/Roel/darktable/build-hr/bin -isystem C:/msys64/home/Roel/darktable/src/external -isystem C:/msys64/home/Roel/darktable/src/external/OpenCL -isystem C:/msys64/mingw64/include/glib-2.0 -isystem C:/msys64/mingw64/lib/glib-2.0/include -isystem C:/msys64/mingw64/include/gtk-3.0 -isystem C:/msys64/mingw64/include/pango-1.0 -isystem C:/msys64/mingw64/include/harfbuzz -isystem C:/msys64/mingw64/include/freetype2 -isystem C:/msys64/mingw64/include/libpng16 -isystem C:/msys64/mingw64/include/fribidi -isystem C:/msys64/mingw64/include/cairo -isystem C:/msys64/mingw64/include/lzo -isystem C:/msys64/mingw64/include/pixman-1 -isystem C:/msys64/mingw64/include/gdk-pixbuf-2.0 -isystem C:/msys64/mingw64/include/atk-1.0 -isystem C:/msys64/mingw64/include/libxml2 -isystem C:/msys64/mingw64/include/libsoup-2.4 -isystem C:/msys64/mingw64/include/OpenEXR -isystem C:/msys64/mingw64/include/lensfun -isystem C:/msys64/mingw64/include/librsvg-2.0 -isystem C:/msys64/mingw64/include/json-glib-1.0 -isystem C:/msys64/mingw64/include/openjpeg-2.4 -isystem C:/msys64/mingw64/include/libsecret-1 -isystem C:/msys64/mingw64/include/GraphicsMagick -isystem C:/msys64/mingw64/include/osmgpsmap-1.0 -Wall -Wno-format -Wshadow -Wtype-limits -Wvla -Wold-style-declaration -Wno-unknown-pragmas -Wno-error=varargs -Wno-format-truncation -Wno-error=address-of-packed-member -std=c99 -fopenmp -march=native -msse2 -g -mfpmath=sse -O3 -DNDEBUG -O3 -ffast-math -fno-finite-math-only -fexpensive-optimizations -fvisibility=hidden   -DUNICODE -D_UNICODE -Werror -Wfatal-errors -include common/module_api.h -include iop/iop_api.h -MD -MT lib/darktable/plugins/CMakeFiles/highlights.dir/introspection_highlights.c.obj -MF lib\darktable\plugins\CMakeFiles\highlights.dir\introspection_highlights.c.obj.d -o lib/darktable/plugins/CMakeFiles/highlights.dir/introspection_highlights.c.obj -c C:/msys64/home/Roel/darktable/build-hr/lib/darktable/plugins/introspection_highlights.c
In file included from C:/msys64/home/Roel/darktable/build-hr/lib/darktable/plugins/introspection_highlights.c:60:
C:/msys64/home/Roel/darktable/src/iop/highlights.c: In function 'heat_PDE_diffusion':
C:/msys64/home/Roel/darktable/src/iop/highlights.c:1256:31: error: 'B_SPLINE_TO_LAPLACIAN_2' undeclared (first use in this function)
 1256 |           covariance_HF[c] += B_SPLINE_TO_LAPLACIAN_2 * (neighbour_pixel_HF[k][c] - means_HF[c]) * (neighbour_pixel_HF[k][guiding_channel] - means_HF[guiding_channel]) / 9.f;
      |                               ^~~~~~~~~~~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
```

Any idea for a fix? I'm building from an empty directory, as is normal.

aurelienpierre commented:
> As we are at it, why the highlight recovery module default to "clip"? Can't we use a safe default like "reconstruct in LCh"?

It's really not safe.

> Hi @aurelienpierre I liked to try this out, but I get compilation errors on Windows MSYS2:

Fixed by 99f9958, I forgot to commit one file where that constant is defined.

TurboGit commented:
> It's really not safe.

Can be worst than "clip" :)

aurelienpierre commented:
> Can be worst than "clip" :)

Clip is the only white-balance-agnostic method. All the others rely on the quality of the WB… which is not guaranteed.

jenshannoschwalm commented:
@aurelienpierre don't know if #10716 has a chance to get into dt codebase too but if we bumb the module version it might be good to leave room in modules parameters for lets say two ints and floats each. Wouldn't hurt but would make testing for other folks much easier.

Or maybe do a seperate pr just handling the module version to prepare for such a scenario?

aurelienpierre commented:
@jenshannoschwalm actually @rawfiner raised a good point : why are we taking the trouble of recovering HL before demosaicing ? We need as many branches as CFA patterns and a test demosaicing that will be ditched.

So I wonder if this wouldn't be better in a new module after demosaicing.

jenshannoschwalm commented:
For me this seems to be good too. Would also

- allow non-bayer sensors
- defer the green channel handling completely to the demosaicer ...

Would offer to implement the modules framework ...

aurelienpierre commented:
Great, I'll take it. I you reserve me 3 floats in addition of the clipping threshold in the params structure, then I can rebase my work on yours without legacy import.

rawfiner commented:
To sum up what I *think* between being before or after demosaic for highlight reconstruction:

|                                   | before demosaic                    | after demosaic                                                                                                                                                                                                                         |
|-----------------------------------|------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| performance                       | no downscaling in darkroom         | downscaling in darkroom helps for preview performance                                                                                                                                                                                                                |
| preview accuracy                  | perfect (no downscaling)           | maybe imperfect  (need to generate previews from a downscaled image, and downscaling may make some pixel value go below the threshold)                                                                                                                                                                                                                       |
| detect pixels to correct          | straightforward                    | maybe need to select all pixels > threshold, and then extend the selection to the surrounding ones to take into account the fact that demosaic combined invalid values with valid ones, making invalid values lower than the threshold |
| code maintenance / camera support | may not be easy to support all CFA | camera support is straightforward. Probably a bit less code, and less complicated.                                                                                                                                                     |

I think it is worth trying to have the module after demosaic for both of your approaches:
- for @aurelienpierre 's approach, it would help a lot for preview performance
- I would like a lot being able to @jenshannoschwalm approach as well on my fuji files with xtrans CFA :-)

aurelienpierre commented:
Notice that I already postfilter the clipping mask with a 5x5 box blur because the reconstruction creates overshoots otherwise.

jenshannoschwalm commented:
On second thought , processing per channel will be 4-fold, correct?

aurelienpierre commented:
> On second thought , processing per channel will be 4-fold, correct?

what do you mean ?

jenshannoschwalm commented:
At least for the segmentation I do, in Bayer space the size for width / height are half size compared to after demosaicing.

rawfiner commented:
Indeed if you did segmentation using red or blue channel, you get 4x more pixels after demosaic, so it can result in a slowdown

aurelienpierre commented:
You can always downsample though.

aurelienpierre commented:
No longer WIP and ready for review for the CPU path.

aurelienpierre commented:
@jenshannoschwalm I have some serious work to edit (show photography with stage lights) for which I would like to use this, albeit unmerged, without introducing history mess-ups later. Do you mind if we bump your work to a v3 of the module params ?

jenshannoschwalm commented:
> Do you mind if we bump your work to a v4 of the module params or do you want me to include your params as part of v4 here ?

I would actually prefer to have the parameters for my algorithm to be included in your version bump too, would make it easier to test & use both algorithms atm without later hassle. 3 floats are fine as in

```
  float reconstructing;    // $MIN: 0.0 $MAX: 1.0  $DEFAULT: 0.4 $DESCRIPTION: "cast balance"
  float combine;           // $MIN: 0.0 $MAX: 10.0 $DEFAULT: 2.0 $DESCRIPTION: "combine segments"
  float synthesis;

```

ChristianBirzer commented:
Hi,
I just tried to test your branch but get an segfault as soon as I switch the highlight recovery module to guided laplacian.
![grafik](https://user-images.githubusercontent.com/96134139/148549594-a449a3bb-0ec0-4a47-815d-58a80fd885e6.png)
Crash in file bspline.h, function spars_scalar_product. Seems the indices array has weird values.

Two frames back in the call stack:
![grafik](https://user-images.githubusercontent.com/96134139/148550970-ea19e4fc-bd42-4985-be2a-c3a4929d8d26.png)


I'm not sure if this crash is relevant at the current stage. Just wanted to mention it, maybe its useful :-)

System: Win10, compiled from commit edded50f with msys.

Let me know if there is anything I can do to help.

aurelienpierre commented:
@ChristianBirzer `mult` should be between 1 and 1024, in your case it's 5360 so that's probably the cause. But I don't see how it can be set to a such value, and this code is already used in several places in dt (diffuse or sharpen, filmic reconstruction), so it's weird that it would crash only here. Can you run the program in terminal and print the stdout ? That should tell us how many wavelets scales are processed, which may explain how `mult` ends up that high.

ChristianBirzer commented:
Here is the console output as generated with the -d all option. Beginning of the output is after switching to darkroom module and just before setting the highlight recovery to guided laplacian.
Only removed hundreds of Pango font errors...

I cannot see any useful information here but I even don't know what to look at 🙄

[log.txt](https://github.com/darktable-org/darktable/files/7829531/log.txt)

The strange thing: When breaking in highlights.c, function wavelets_process() the parameter mult is 1 in the call of decompose_2D_Bspline. Width and height have also plausible values. padded_size is 5360. 
After stepping into the decompose_2D_Bspline most parameter values are completely confused and the 5360 gets shifted to the mult parameter. 
Ah, I see. I guess the compiler tries to tell me something:
[build] F:/Users/Christian/Dokumente/Entwicklung/OpenSource/Darktable/darktable/src/iop/highlights.c:1419:5: note: the ABI for passing parameters with 64-byte alignment has changed in GCC 4.6
[build]  1419 |     decompose_2D_Bspline(buffer_in, HF[s], buffer_out, width, height, mult, tempbuf, padded_size);

Thought I did a full rebuild with deleting the build directory...

aurelienpierre commented:
You have only one scale processed (s = 0), so `mult` should be 1.

> Thought I did a full rebuild with deleting the build directory...

So, if you restart from scratch, does it fix this issue ?

ChristianBirzer commented:
> So, if you restart from scratch, does it fix this issue ?

Sorry, but no. Did a fresh clone to a new directory and build everything from scratch. Same issue. 
Presumably it's just my build environment. I'll try to do more tests the next days.

ChristianBirzer commented:
The crash as well as the ABI compiler warning only occurs, when I build the debug target. 
In release build, I can select the guide laplacians method. The sliders look different to the screenshots above, but I don't know if this is ok or not. When switching to guide laplacian, I can see a difference in the output image though. 

![grafik](https://user-images.githubusercontent.com/96134139/148642005-6ea97207-64b0-4dc5-992d-6679a353b030.png)

jenshannoschwalm commented:
Some feedback:
1. noisy debugging output "Wavelet decompose : scale x"
2. I thought my machine is not so bad :-) 45Mpix images taking ~16secs on a 8C/16T thing -march=native
3. Strange results for some images like this (watch for the footpath in the front or spots on the grass)
![Bildschirmfoto von 2022-01-08 18-11-37](https://user-images.githubusercontent.com/50982232/148653302-fbe31016-16ca-4671-804b-7d2e2e7e0813.png)

elstoc commented:
Haven't tested this but looking at the last screenshot, is "guide laplacians" the right name for the method. Should it not be something "guided laplacians"?

aurelienpierre commented:
> Should it not be "guided laplacians"?

All the other options use verbs (clip, reconstruct) so I chose consistency. I don't care either ways.

> noisy debugging output "Wavelet decompose : scale x"

For the time being, that helped with Christian's issue above.

> I thought my machine is not so bad :-) 45Mpix images taking ~16secs on a 8C/16T thing -march=native

Well… 2 wavelets decompositions with internal anisotropic laplacian filter and guided filter to compute at each scale… That should get better with OpenCL.

> Strange results for some images like this (watch for the footpath in the front or spots on the grass)

That seems to be a misconfig of filmic white because I'm unable to reproduce:
![Screenshot_20220109_160024](https://user-images.githubusercontent.com/2779157/148687928-00571e56-691f-4a35-b377-a693c27396ae.jpg)

elstoc commented:
> All the other options use verbs (clip, reconstruct) so I chose consistency. I don't care either ways.

Hmm. I've always read these as "clip \[highlights\]" and "reconstruct \[highlights\]" but "guide laplacians highlights" doesn't really fit this pattern. "guide" might make sense in this context but "guide laplacians" doesn't because the thing being acted on by the verb is different. If rewritten as suggested I would read it as "\[use\] guided laplacians", with the verb being the implied "use".

MStraeten commented on src/iop/highlights.c:
```diff
@@ -1151,12 +1151,32 @@ static inline void guide_laplacians(const float *const restrict high_freq, const
+          if(variance_HF[c] > guiding_value_HF)
           {
             // Find the channel least likely to be clipped
             guiding_value = white_unbalanced_color;
```
guiding_value as well as white_unbalanced_color are undeclared here
```
 65%] Building C object lib/darktable/plugins/CMakeFiles/highlights.dir/introspection_highlights.c.o
In file included from /Users/martinstraeten/src/darktable/build/lib/darktable/plugins/introspection_highlights.c:65:
/Users/martinstraeten/src/darktable/src/iop/highlights.c:1182:13: fatal error: use of undeclared identifier 'guiding_value'
            guiding_value = white_unbalanced_color;
            ^
1 error generated.
```

matze commented:
No strange behavior on AMD although it is a bit slow.

aurelienpierre commented:
I know it's slow, but I don't see any workaround…

aurelienpierre commented:
Note: I have changed the sequence a bit, for the sake of run time.

The number of wavelets scales is limited to 8, meaning a max radius of 256 px. But since the process is now iterative, we need to double the padding size to avoid straight artifacts at tile's edges. So the max padding is 512 px.

The result is very large blown areas, with radii larger than 512 px will not be corrected at all. It's either this or insane runtimes.

SoupyGit commented:
"The result is very large blown areas, with radii larger than 512 px will not be corrected at all. It's either this or insane runtimes."

Just a thought, could potentially have both, with a user drop down to select blown areas: small or large, where small allows for many iterations, but large allows only for one.

MStraeten commented:
Build fails on win10:
```
In file included from C:/msys64/home/me/darktable/build/lib/darktable/plugins/introspection_highlights.c:71:
C:/msys64/home/SHM/darktable/src/iop/highlights.c:1564: error: unterminated #if
 1564 | #if HAVE_OPENCL
      |
```
introduced with https://github.com/darktable-org/darktable/pull/10711/commits/dec805c1d75ae7c25361e29494e0f9f7016a4b52

aurelienpierre commented:
I solved conflicts, rebased and fixed the anisotropic diffusion kernel similarly to what @flannelhead did for diffuse.c.

@TurboGit  this is ready to go !

TurboGit commented:
I've not looked at the code yet, testing it I see that I have some magenta color cast on tow test images.

I'm using Filmic and recontruct is threshold is negative -0.8IL.

First image, module **highlight reconstruction** set to

1. clip
![image](https://user-images.githubusercontent.com/467069/161285719-9412df09-54ee-4818-ae0b-27205192e9eb.png)

2. LCh
![image](https://user-images.githubusercontent.com/467069/161286115-09c15ec4-7257-48d8-b469-8e467159e134.png)

3. color
![image](https://user-images.githubusercontent.com/467069/161286208-a6c0ea5d-81e1-4c64-b3d2-0c77fecb82fa.png)

4. guider laplacian
![image](https://user-images.githubusercontent.com/467069/161286268-e0d14925-0341-42ac-a758-38e8d4123bac.png)

Second image:

1. clip
![image](https://user-images.githubusercontent.com/467069/161286493-25ac2c4a-0c93-4490-97c0-0b290a6c678d.png)

2. LCh
![image](https://user-images.githubusercontent.com/467069/161286546-f0f88af5-4d6a-4f47-bbf4-f4e816a8b661.png)

3. color
![image](https://user-images.githubusercontent.com/467069/161286573-cc776cbb-bf74-4b89-a3e1-1b273d4adbca.png)

4. guider laplacian
![image](https://user-images.githubusercontent.com/467069/161286633-0e56b965-3b44-4924-8a07-345025620daf.png)

Is that a wrong use on my side?

Also I see that the the X-Trans (RAF) are not supported, I suppose this is expected?

TurboGit commented:
If I change the white rel exposure in Filmic it somehow solves the issue. To me this method seems to be very sensitive to while being too strong, again not sure it is expected. Until the doc is there it is hard for me to guess all that :)

aurelienpierre commented:
@TurboGit It's important to understand that this algo doesn't care about colors. There is no desaturation or no explicit color propagation like the other methods do. What is done is:

1. attempt to copy-paste details (modelled as the laplacian) from valid channels to clipped channels,
1. attempt to diffuse the content from neighbouring regions.

This means that the algo will have an hard time with large areas, for which you need to increase the diameter of reconstruction and the number of iterations, at the expense of runtime.

Also, since filmic v6 does not desaturate highlights mandatorily, obviously magenta casts are more visible.

TurboGit commented:
> @TurboGit It's important to understand that this algo doesn't care about colors. There is no desaturation or no explicit color propagation like the other methods do. What is done is:
> 
>     1. attempt to copy-paste details (modelled as the laplacian) from valid channels to clipped channels,
> 
>     2. attempt to diffuse the content from neighbouring regions.
> 
> 
> This means that the algo will have an hard time with large areas, for which you need to increase the diameter of reconstruction and the number of iterations, at the expense of runtime.
> 
> Also, since filmic v6 does not desaturate highlights mandatorily, obviously magenta casts are more visible.

Ok, will test again with that in mind. Thanks for the explanation!

TurboGit commented:
I have hard time to find a picture where it makes a difference with the other algorithm. I generally find that "color reconstruction" works best without the color cast and that Laplacian gives almost the same result but with noise added with the image is high iso.

@aurelienpierre : Can you share a picture where this algorithm gives results that cannot be achieved with the others?

TurboGit commented:
There is a missing update somewhere. To reproduce:
1. select an image
2. in highlight reconstruction select laplacian
3. using filmstrip change image
4. see that the highlight reconstruction is still laplacian
5. try to click on iteration slider
6. see error, update is now done and the slider has vanished

TurboGit reviewed (changes requested):
GUI update bug.

aurelienpierre commented:
@TurboGit GUI bug fixed, it was an easy one.

Next commit changes slightly the reconstruction logic to fix an artifact I uncovered. It actually removes some operation.

It will allow to rewrite the wavelets decomposition to be sequential instead of parallel, meaning only 3 buffers (LF, HF, synthesis) will be needed instead of having to store all HF buffers. That will prevent tiling and the huge perf penalty that goes with it.

TurboGit reviewed (approved):
Looks ok, thanks!
